adapters/http-vertx-base/src/main/java/org/eclipse/hono/adapter/http/AbstractVertxBasedHttpProtocolAdapter.java

```diff
@@ -975,24 +974,34 @@ protected void setNonEmptyResponsePayload(final HttpServerResponse response, fin
             }
             return Future.succeededFuture();
         }
+        uploadMessageSpan.setTag(MessageHelper.APP_PROPERTY_DEVICE_TTD, ttdSecs);
+
+        final Span waitForCommandSpan = TracingHelper
+                .buildChildSpan(tracer, uploadMessageSpan.context(),
+                        "wait for command", getTypeName())
+                .withTag(Tags.SPAN_KIND.getKey(), Tags.SPAN_KIND_CLIENT)
+                .withTag(TracingHelper.TAG_TENANT_ID, tenantObject.getTenantId())
+                .withTag(TracingHelper.TAG_DEVICE_ID, deviceId)
+                .start();
 
-        currentSpan.setTag(MessageHelper.APP_PROPERTY_DEVICE_TTD, ttdSecs);
         final Handler<CommandContext> commandHandler = commandContext -> {
 
             Tags.COMPONENT.set(commandContext.getTracingSpan(), getTypeName());
             final Command command = commandContext.getCommand();
             final Sample commandSample = getMetrics().startTimer();
-            if (isCommandValid(command, currentSpan)) {
+            if (isCommandValid(command, waitForCommandSpan)) {
 
                 if (requestProcessed.compareAndSet(false, true)) {
-                    checkMessageLimit(tenantObject, command.getPayloadSize(), currentSpan.context())
+                    CommandConsumer.logReceivedCommandToSpan(command, waitForCommandSpan);
+                    checkMessageLimit(tenantObject, command.getPayloadSize(), waitForCommandSpan.context())
                     .onComplete(result -> {
                         if (result.succeeded()) {
                             addMicrometerSample(commandContext, commandSample);
                             // put command context to routing context and notify
                             ctx.put(CommandContext.KEY_COMMAND_CONTEXT, commandContext);
                         } else {
                             commandContext.reject(getErrorCondition(result.cause()));
+                            waitForCommandSpan.log("rejected command for device: " + result.cause());
```

Comment:
FMPOV the cause (Throwable) should be logged here as intended by the OpenTracing spec, i.e. using *Fields.ERROR_KIND* and *Fields.ERROR_OBJECT* to be consistent with other cases where we use *TracingHelper.log(Throwable)*. We might want to consider adding *Fields.STACK* to that method as well ...

Reply:
I think the stack is not important here. The log message is supposed to contain `result.cause().toString()`.

Concerning the "ERROR_KIND" and "ERROR_OBJECT" fields: These would of course be set if  `Tracing.logError` was used here. I didn't want to use this because I didn't want the HTTP PUT span to be marked as an error span in this case (that wasn't done so before).

But thinking again, I think this case should indeed be marked as an error span, reflecting the "check message limit failed" error. I have adapted the commit accordingly, also setting an error for the "received invalid command message" case.